python/cudf/cudf/utils/utils.py

```diff
@@ -371,3 +377,45 @@ def _cudf_nvtx_annotate(func, domain="cudf_python"):
 _dask_cudf_nvtx_annotate = partial(
     _cudf_nvtx_annotate, domain="dask_cudf_python"
 )
+
+
+class CachedInstanceMeta(type):
+    """
+    Metaclass that caches `maxsize` instances.
+    """
+
+    def __new__(self, names, bases, attrs, maxsize=128):
+        self.__maxsize = maxsize
+        return type.__new__(self, names, bases, attrs)
+
+    def __init__(self, names, bases, attrs, **kwargs):
+        self.__instances = OrderedDict()
+
+    def __call__(self, *args, **kwargs):
+        # the cache key is constructed from args and kwargs, and also
+        # the _types_ of args and kwargs, since objects of different
+        # types can compare equal
+        arg_tuple = (
+            args
+            + tuple(kwargs.values())
+            + tuple(map(type, args))
+            + tuple(map(type, kwargs.values()))
+        )
+        try:
+            # try retrieving an instance from the cache:
+            self.__instances.move_to_end(arg_tuple)
```

Comment:
I saw the conversation about changing to `dict` above, which I support. If the sole reason for using `OrderedDict` is for the `move_to_end` method, I would suggest using the snippet recommended here with a comment to indicate the intent: https://docs.python.org/3/library/collections.html#ordereddict-objects

> A regular [dict](https://docs.python.org/3/library/stdtypes.html#dict) can emulate `OrderedDict`’s `od.move_to_end(k, last=True)` with `d[k] = d.pop(k)` which will move the key and its associated value to the rightmost (last) position.

The `dict` is a faster class overall, and we should try to use it if possible.

Reply:
Those docs specifically call out that OrderedDicts are suitable for implementing LRU caches though.

Reply:
The docs do mention that, but `OrderedDict` is still not the most efficient tool for that job. I wrote a microbenchmark here that demonstrates how `dict` is more efficient, even when it does a bit more work: https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/dict_lru.py
```
       dict: 0.00619369
OrderedDict: 0.00837117
   UserDict: 0.021145
```
In the cases I've seen and benchmarked in the past, `OrderedDict` is not as fast and adds an unnecessary import to the code. (Deriving from `UserDict` is vastly slower than owning a `dict`.)

Reply:
The outcome here is not important to me -- 25% perf on a cache isn't a big deal. Just wanted to make sure we are considering the extra import and slower perf, versus the benefit of having a slightly more clear `move_to_end` method.

Reply:
Using `dict` here is actually not as obviously a win as you might think (microbenchmarking is hard)

```
In [1]: %%timeit
   ...: for i in range(1000):
   ...:     x = ScalarDict(i % 128, dtype=np.float64)
   ...: 
663 µs ± 2.25 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [2]: %%timeit
   ...: for i in range(1000):
   ...:     x = ScalarOrderedDict(i % 128, dtype=np.float64)
   ...: 
661 µs ± 1.64 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

```

A _much_ bigger win comes from making this cache _sui generis_, rather than a general implementation (since most of the cost is actually in constructing the key and doing the argument packing/unpacking:
```
In [3]: %%timeit
   ...: for i in range(1000):
   ...:     x = ScalarDictSpecific(i % 128, dtype=np.float64)
   ...: 
282 µs ± 875 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [4]: %%timeit
   ...: for i in range(1000):
   ...:     x = ScalarOrderedDictSpecific(i % 128, dtype=np.float64)
   ...: 
277 µs ± 1.74 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

```

So, if we care about the cache lookup performance, and don't anticipate needing to cache many other things, it's probably better to make this caching specific to `Scalar` and avoid the splat and splat-splat overhead.

<details><summary>Code</summary>

```python
from collections import OrderedDict

import numpy as np


class CacheOrderedDict(type):
    """
    Metaclass that caches `maxsize` instances.
    """

    def __new__(self, names, bases, attrs, maxsize=128):
        self.__maxsize = maxsize
        return type.__new__(self, names, bases, attrs)

    def __init__(self, names, bases, attrs, **kwargs):
        self.__instances = OrderedDict()

    def __call__(self, *args, **kwargs):
        # the cache key is constructed from args and kwargs, and also
        # the _types_ of args and kwargs, since objects of different
        # types can compare equal
        arg_tuple = (
            args +
            tuple(kwargs.items()) +
            tuple(map(type, args)) +
            tuple(map(type, kwargs.values()))
        )
        try:
            # try retrieving an instance from the cache:
            self.__instances.move_to_end(arg_tuple)
            return self.__instances[arg_tuple]
        except KeyError:
            # if an instance couldn't be found in the cache,
            # construct it and add to cache:
            obj = super().__call__(*args, **kwargs)
            self.__instances[arg_tuple] = obj
            if len(self.__instances) > self.__maxsize:
                self.__instances.popitem(last=False)
            return obj
        except TypeError:
            # couldn't hash args/kwargs, don't cache:
            return super().__call__(*args, **kwargs)

class CacheDict(type):
    """
    Metaclass that caches `maxsize` instances.
    """

    def __new__(self, names, bases, attrs, maxsize=128):
        self.__maxsize = maxsize
        return type.__new__(self, names, bases, attrs)

    def __init__(self, names, bases, attrs, **kwargs):
        self.__instances = dict()

    def __call__(self, *args, **kwargs):
        # the cache key is constructed from args and kwargs, and also
        # the _types_ of args and kwargs, since objects of different
        # types can compare equal
        arg_tuple = (
            args +
            tuple(kwargs.items()) +
            tuple(map(type, args)) +
            tuple(map(type, kwargs.values()))
        )
        try:
            # try retrieving an instance from the cache:
            val = self.__instances.pop(arg_tuple)
            self.__instances[arg_tuple] = val
            return val
        except KeyError:
            # if an instance couldn't be found in the cache,
            # construct it and add to cache:
            obj = super().__call__(*args, **kwargs)
            self.__instances[arg_tuple] = obj
            if len(self.__instances) > self.__maxsize:
                k = next(iter(self.__instances))
                self.__instances.pop(k)
            return obj
        except TypeError:
            # couldn't hash args/kwargs, don't cache:
            return super().__call__(*args, **kwargs)


class CacheOrderedDictSpecific(type):
    """
    Metaclass that caches `maxsize` instances.
    """

    def __new__(self, names, bases, attrs, maxsize=128):
        self.__maxsize = maxsize
        return type.__new__(self, names, bases, attrs)

    def __init__(self, names, bases, attrs, **kwargs):
        self.__instances = OrderedDict()

    def __call__(self, value, dtype=None):
        # the cache key is constructed from args and kwargs, and also
        # the _types_ of args and kwargs, since objects of different
        # types can compare equal
        arg_tuple = (value, dtype, type(value), type(dtype))
        try:
            # try retrieving an instance from the cache:
            self.__instances.move_to_end(arg_tuple)
            return self.__instances[arg_tuple]
        except KeyError:
            # if an instance couldn't be found in the cache,
            # construct it and add to cache:
            obj = super().__call__(value, dtype=dtype)
            self.__instances[arg_tuple] = obj
            if len(self.__instances) > self.__maxsize:
                self.__instances.popitem(last=False)
            return obj
        except TypeError:
            # couldn't hash args/kwargs, don't cache:
            return super().__call__(value, dtype=dtype)


class CacheDictSpecific(type):
    """
    Metaclass that caches `maxsize` instances.
    """

    def __new__(self, names, bases, attrs, maxsize=128):
        self.__maxsize = maxsize
        return type.__new__(self, names, bases, attrs)

    def __init__(self, names, bases, attrs, **kwargs):
        self.__instances = dict()

    def __call__(self, value, dtype=None):
        # the cache key is constructed from args and kwargs, and also
        # the _types_ of args and kwargs, since objects of different
        # types can compare equal
        arg_tuple = (value, dtype, type(value), type(dtype))
        try:
            # try retrieving an instance from the cache:
            val = self.__instances.pop(arg_tuple)
            self.__instances[arg_tuple] = val
            return val
        except KeyError:
            # if an instance couldn't be found in the cache,
            # construct it and add to cache:
            obj = super().__call__(value, dtype=dtype)
            self.__instances[arg_tuple] = obj
            if len(self.__instances) > self.__maxsize:
                k = next(iter(self.__instances))
                self.__instances.pop(k)
            return obj
        except TypeError:
            # couldn't hash args/kwargs, don't cache:
            return super().__call__(value, dtype=dtype)


class Uncached:
    def __init__(self, value, dtype=None):
        self.value = value
        self.dtype = dtype


class ScalarOrderedDict(Uncached, metaclass=CacheOrderedDict):
    pass


class ScalarDict(Uncached, metaclass=CacheDict):
    pass


class ScalarOrderedDictSpecific(Uncached, metaclass=CacheOrderedDictSpecific):
    pass


class ScalarDictSpecific(Uncached, metaclass=CacheDictSpecific):
    pass
```

</details>

Reply:
> dict is more efficient, even when it does a bit more work: https://github.com/bdice/python_microbenchmarks/blob/main/microbenchmarks/dict_lru.py

Thanks! That is interesting.

Perhaps some of the difference can be accounted for by the fact that `LastUpdatedOrderedDict` uses inheritance where `LastUpdatedDict` uses composition? Skipping the inheritance in `LastUpdatedOrderedDict`, I see the difference is less dramatic:

```python
class LastUpdatedOrderedDict():
    'Store items in the order the keys were last added'

    def __init__(self):
        self._cache = OrderedDict()

    def __getitem__(self, key):
        return self._cache[key]

    def __setitem__(self, key, value):
        self._cache[key] = value
        self._cache.move_to_end(key)
```

```
              dict: 0.00523313
       OrderedDict: 0.00548418
          UserDict: 0.016895
```

Reply:
> So, if we care about the cache lookup performance, and don't anticipate needing to cache many other things, it's probably better to make this caching specific to Scalar and avoid the splat and splat-splat overhead.

Admittedly, I can't think of many other uses currently. [cudf.DateOffset](https://github.com/rapidsai/cudf/blob/branch-22.10/python/cudf/cudf/core/tools/datetimes.py#L379e) comes to mind, but it is itself just a collection of `Scalar` objects.

Although I personally prefer the separation of concerns between caching and `Scalar`, I'm open to embedding the logic into the `Scalar` class given the perf difference. I'd let @bdice break the tie here :)

Reply:
Fascinating analysis! It’s tough to be the tie breaking vote here. 😆 I favor the scalar specific implementation because we don’t have another great use case in mind. We won’t be caching columns on the host or anything like that.

Reply:
Fix'd:

- @wence-'s benchmmarks show that `OrderedDict` and `dict` have roughly the same perf in this context, so I kept the use of `OrderedDict`.  The use of the `popitem()` and `move_to_end()` methods is more explicit than the equivalent `dict` methods. I'm not super concerned about the additional `import`.

- I replaced the use of `CachedInstanceMeta`,  with a more scpecific  `CachedScalarInstanceMeta`. I considered not using a metaclass at all and embedding the logic in `Scalar` directly, but I decided against that in case we (1) want to generalize caching to other types, or (2) decide for some reason that caching was a bad idea.